backend/tests/e2e/parallel_test_base.py

```
@@ -0,0 +1,433 @@
+            test_client, max_wait_seconds=max_wait_seconds
+        )
+        
+        # If session failed, get detailed error info before asserting
+        if final_status == "failed":
+            try:
+                detail_response = test_client.get(f"/api/v1/history/sessions/{session_id}")
+                if detail_response.status_code == 200:
+                    detail_data = detail_response.json()
+                    error_message = detail_data.get("error_message", "No error message")
+                    print(f"❌ Session failed with error: {error_message}")
+                    assert False, f"Session failed with status: {final_status}, error: {error_message}"
```

Ruff (0.14.8)

Do not `assert False` (`python -O` removes these calls), raise `AssertionError()`

Replace `assert False`

(B011)